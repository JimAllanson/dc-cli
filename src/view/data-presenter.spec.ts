import DataPresenter from './data-presenter';

const stdoutWriteSpy = jest.spyOn(process.stdout, 'write');

describe('DataPresenter', (): void => {
  const argv = {
    $0: 'test',
    _: ['test'],
    clientId: 'client-id',
    clientSecret: 'client-id',
    hubId: 'hub-id'
  };

  beforeAll(() => {
    stdoutWriteSpy.mockImplementation(() => true);
  });

  afterAll(() => {
    jest.restoreAllMocks();
  });

  it('should render a vertical table', (): void => {
    const data = {
      toJSON: (): { [key: string]: string } => ({
        foo: 'bar',
        key: 'value'
      })
    };
    new DataPresenter(argv, data).render();
    expect(stdoutWriteSpy.mock.calls[0][0]).toMatchSnapshot();
  });

  it('should render a horizontal table', (): void => {
    const data = {
      toJSON: (): { [key: string]: string }[] => [
        {
          foo: 'bar',
          key: 'value'
        }
      ]
    };
    new DataPresenter(argv, data).render();
    expect(stdoutWriteSpy.mock.calls[0][0]).toMatchSnapshot();
  });

  it('should render a horizontal table with some page information', (): void => {
    const data = {
      toJSON: (): { [key: string]: string }[] => [
        {
          foo: 'bar',
          key: 'value'
        }
      ],
      page: {
        number: 0,
        totalPages: 20
      }
    };
    new DataPresenter(argv, data).render();
    expect(stdoutWriteSpy.mock.calls[1][0]).toMatchSnapshot();
  });

  it('should render some json', (): void => {
    const data = {
      toJSON: (): { [key: string]: string } => ({
        foo: 'bar',
        key: 'value'
      })
    };
    new DataPresenter({ ...argv, json: true }, data).render();
    expect(stdoutWriteSpy.mock.calls[0][0]).toMatchSnapshot();
  });

  it('should parse some data', (): void => {
    const data = {
      foo: 'bar',
      key: 'value'
    };

    new DataPresenter(argv, { ...data, toJSON: (): { [key: string]: string } => data })
      .parse(({ foo }): { [key: string]: string } => ({ foo }))
      .render();
    expect(stdoutWriteSpy.mock.calls[0][0]).toMatchSnapshot();
  });
});
