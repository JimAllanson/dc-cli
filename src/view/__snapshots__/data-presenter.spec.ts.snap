// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataPresenter collection of items should render a collection of items as json 1`] = `
"[
  {
    \\"foo\\": \\"bar1\\",
    \\"key\\": \\"value1 value1 value1 value1 value1 value1 value1 value1 value1 value1 value1 value1 value1 value1 value1 value1 value1 value1 value1 value1 \\"
  },
  {
    \\"foo\\": \\"bar2\\",
    \\"key\\": \\"value2 value2 value2 value2 value2 value2 value2 value2 value2 value2 value2 value2 value2 value2 value2 value2 value2 value2 value2 value2 \\"
  }
]"
`;

exports[`DataPresenter collection of items should render a collection of items horizontal table using with table config 1`] = `
"+------+------------------------------------------------------------------------------------------------------+
| [1mfoo[22m  | [1mkey[22m                                                                                                  |
|------|------------------------------------------------------------------------------------------------------|
| bar1 | value1 value1 value1 value1 value1 value1 value1 value1 value1 value1 value1 value1 value1 value1 va |
|      | lue1 value1 value1 value1 value1 value1                                                              |
|------|------------------------------------------------------------------------------------------------------|
| bar2 | value2 value2 value2 value2 value2 value2 value2 value2 value2 value2 value2 value2 value2 value2 va |
|      | lue2 value2 value2 value2 value2 value2                                                              |
+------+------------------------------------------------------------------------------------------------------+

"
`;

exports[`DataPresenter collection of items should render a collection of items horizontal table with some page information 1`] = `
"+------+----------------------------------------------------------------------------------------------------------------------------------------------+
| [1mfoo[22m  | [1mkey[22m                                                                                                                                          |
|------|----------------------------------------------------------------------------------------------------------------------------------------------|
| bar1 | value1 value1 value1 value1 value1 value1 value1 value1 value1 value1 value1 value1 value1 value1 value1 value1 value1 value1 value1 value1  |
|------|----------------------------------------------------------------------------------------------------------------------------------------------|
| bar2 | value2 value2 value2 value2 value2 value2 value2 value2 value2 value2 value2 value2 value2 value2 value2 value2 value2 value2 value2 value2  |
+------+----------------------------------------------------------------------------------------------------------------------------------------------+

[1mDisplaying page 1 of 20[22m
"
`;

exports[`DataPresenter collection of items should render a collection of items in a horizontal table 1`] = `
"+------+----------------------------------------------------------------------------------------------------------------------------------------------+
| [1mfoo[22m  | [1mkey[22m                                                                                                                                          |
|------|----------------------------------------------------------------------------------------------------------------------------------------------|
| bar1 | value1 value1 value1 value1 value1 value1 value1 value1 value1 value1 value1 value1 value1 value1 value1 value1 value1 value1 value1 value1  |
|------|----------------------------------------------------------------------------------------------------------------------------------------------|
| bar2 | value2 value2 value2 value2 value2 value2 value2 value2 value2 value2 value2 value2 value2 value2 value2 value2 value2 value2 value2 value2  |
+------+----------------------------------------------------------------------------------------------------------------------------------------------+

"
`;

exports[`DataPresenter collection of items should render using the supplied map() for a collection of items in a horizontal table 1`] = `
"+------+
| [1mfoo[22m  |
|------|
| bar1 |
|------|
| bar2 |
+------+

"
`;

exports[`DataPresenter single item should not output page information for a single item 1`] = `
"+----------+----------------------------------------------------------------------------------------------------------------------------+
| [1mProperty[22m | [1mValue[22m                                                                                                                      |
|----------|----------------------------------------------------------------------------------------------------------------------------|
| foo      | \\"bar\\"                                                                                                                      |
|----------|----------------------------------------------------------------------------------------------------------------------------|
| key      | \\"value value value value value value value value value value value value value value value value value value value value \\" |
+----------+----------------------------------------------------------------------------------------------------------------------------+

"
`;

exports[`DataPresenter single item should render a single item as json 1`] = `
"{
  \\"foo\\": \\"bar\\",
  \\"key\\": \\"value value value value value value value value value value value value value value value value value value value value \\"
}"
`;

exports[`DataPresenter single item should render a single item as vertical table using tableUserConfig 1`] = `
"+----------+------------------------------------------------------------------------------------------------------+
| [1mProperty[22m | [1mValue[22m                                                                                                |
|----------|------------------------------------------------------------------------------------------------------|
| foo      | \\"bar\\"                                                                                                |
|----------|------------------------------------------------------------------------------------------------------|
| key      | \\"value value value value value value value value value value value value value value value value val |
|          | ue value value value \\"                                                                               |
+----------+------------------------------------------------------------------------------------------------------+

"
`;

exports[`DataPresenter single item should render a single item vertical table 1`] = `
"+----------+----------------------------------------------------------------------------------------------------------------------------+
| [1mProperty[22m | [1mValue[22m                                                                                                                      |
|----------|----------------------------------------------------------------------------------------------------------------------------|
| foo      | \\"bar\\"                                                                                                                      |
|----------|----------------------------------------------------------------------------------------------------------------------------|
| key      | \\"value value value value value value value value value value value value value value value value value value value value \\" |
+----------+----------------------------------------------------------------------------------------------------------------------------+

"
`;
